src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// Checksum of data
+    /// </summary>
+    internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
+    {
+        private readonly ImmutableArray<byte> _checkSum;
+        private int? _lazyHash;
+
+        public Checksum(byte[] checksum) :
+            this(ImmutableArray.Create(checksum))
+        {
+        }
+
+        public Checksum(ImmutableArray<byte> checksum)
+        {
+            _lazyHash = null;
+            _checkSum = checksum;
+        }
+
+        public byte[] ToArray()
+        {
+            // TODO: think a way to make this better
+            return _checkSum.ToArray();
+        }
+
+        public bool Equals(Checksum other)
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            if (_checkSum.Length != other._checkSum.Length)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < _checkSum.Length; i++)
+            {
+                if (_checkSum[i] != other._checkSum[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as Checksum);
+        }
+
+        public override int GetHashCode()
+        {
+            if (_lazyHash == null)
+            {
+                // lazily calculate hash for checksum
+                var hash = _checkSum.Length;
+
+                for (var i = 0; i < _checkSum.Length; i++)
+                {
+                    hash = Hash.Combine((int)_checkSum[i], hash);
+                }
+
+                _lazyHash = hash;
+            }
+
+            return _lazyHash.Value;
```

Comment:
No it won't.  You can have interleaved reads and writes and .Net is free to reorder them in the absense of a memory barrier.  For example, in the line: ```_lazyHash = hash;```, the write of the 'boolean' bit on the nullable may happen before the write of the actual value.  Then, another thread may come in and observe that boolean bit being set in ```if (_lazyHash == null)```.  It may then return the unintialized value from teh nullable.